datafusion/physical-expr/src/array_expressions.rs

```diff
@@ -227,16 +227,25 @@ fn compute_array_dims(arr: Option<ArrayRef>) -> Result<Option<Vec<Option<u64>>>>
 }
 
 fn check_datatypes(name: &str, args: &[&ArrayRef]) -> Result<()> {
-    let data_type = args[0].data_type();
-    if !args.iter().all(|arg| {
-        arg.data_type().equals_datatype(data_type)
-            || arg.data_type().equals_datatype(&DataType::Null)
-    }) {
-        let types = args.iter().map(|arg| arg.data_type()).collect::<Vec<_>>();
-        return plan_err!("{name} received incompatible types: '{types:?}'.");
+    let mut data_types = args
+        .iter()
+        .map(|arg| arg.data_type())
+        .collect::<HashSet<_>>();
+    match data_types.len() {
+        1 => Ok(()),
+        2 => {
+            if data_types.remove(&DataType::Null) {
```

Comment:
sorry @Veeupup why we need this `.remove()` ?

Reply:
we want to handle `DataType::Null` in the array function implementation, so we can tolerate `DataType::Null` but can not accept others different types.

Reply:
I can see the null datatype removed from `data_types`, but modified `data_types` never been used after that, this is mostly my concern

Reply:
we do need to really use the modified data_types` because if the 
`data_types.len()` is 2, if one is `Null` we do not need to another data_tpye. 

But if the `data_types` both are not `Null`, which means that input arrays have different `data_types` than we can not handle

Reply:
so `remove` here is just for checking the null datatype is present? would it be more obvious to use `contains` ?
https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.contains

Reply:
you are right, I'll use `contains` instead : )

Reply:
fixed, thanks again!